---
title: "Using EMclassifieR on methylation data"
author: "Jennifer Semple"
date: "7/18/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(EMclassifieR))
```

# The data
DSMF data comes as matrices of reads x positions where each row refers to a single (usually paired-end) read and each column refers to a CpG or GpC position where methylation has been scored. The positions should be in relative coordinates (e.g. -250 to +250 around the TSS), so that matrices from different genes can be compared. 
Some example data is included with the package. First we read in a table with a list of the matrix filenames (including path). The important columns in this table are "filename" and "reads".
```{r}
tablePath="csv/MatrixLog_relCoord_ampTSS.csv"
matTable<-read.csv(system.file("extdata", tablePath, package = "EMclassifieR", mustWork=TRUE),stringsAsFactors=F)
head(matTable)
```

The matrices are in RDS format. For learning it is probably best to process them to remove any reads containing NAs.
```{r}
i=1
data<-readRDS(system.file("extdata", matTable$filename[i], package = "EMclassifieR", mustWork=TRUE))
dim(data)
data<-removeAllNArows(data)
dim(data)
```


## Learning classes for single genes
To look at one gene at a time one can use the minimal matrix which includes only positions with a CpG or GpC motif in them.

```{r,eval=F}
# Configuration
K_range = 2:10                       # Number of classes to be found
maxIterations = 100
convergence_error = 10e-6
numClasses=3
    
outPath="../runResultsEMclass"

set.seed(1)
runEMrepeats(data=data,numClasses=numClasses,convergence_error=convergence_error, maxIterations=maxIterations, repeats=10, outPath=outPath)
    


```

