<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jennifer Semple" />

<meta name="date" content="2020-08-06" />

<title>Using EMclassifieR on methylation data</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Using EMclassifieR on methylation data</h1>
<h4 class="author">Jennifer Semple</h4>
<h4 class="date">2020-08-06</h4>



<div id="the-data" class="section level1">
<h1>The data</h1>
<p>DSMF data comes as matrices of reads x positions where each row refers to a single (usually paired-end) read and each column refers to a CpG or GpC position where methylation has been scored. The positions should be in relative coordinates (e.g. -250 to +250 around the TSS), so that matrices from different genes can be compared. Some example data is included with the package. First we read in a table with a list of the matrix filenames (including path). The important columns in this table are “filename” and “reads”.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>tablePath=<span class="st">&quot;csv/MatrixLog_relCoord_ampTSS.csv&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>matTable&lt;-<span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, tablePath, <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>,</span>
<span id="cb1-3"><a href="#cb1-3"></a>                               <span class="dt">mustWork=</span><span class="ot">TRUE</span>), <span class="dt">stringsAsFactors=</span>F)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw">head</span>(matTable)</span></code></pre></div>
<pre><code>##                                          filename   sample         region reads
## 1 rds/relCoord_ampTSS/dS02-182_WBGene00015947.rds dS02-182 WBGene00015947  1690
## 2 rds/relCoord_ampTSS/dS02-182_WBGene00015955.rds dS02-182 WBGene00015955 10677
## 3 rds/relCoord_ampTSS/dS02-182_WBGene00009620.rds dS02-182 WBGene00009620  3416
## 4 rds/relCoord_ampTSS/dS02-182_WBGene00009621.rds dS02-182 WBGene00009621    31
## 5 rds/relCoord_ampTSS/dS02-182_WBGene00009801.rds dS02-182 WBGene00009801  1177
##   motifs
## 1     25
## 2     36
## 3     26
## 4     25
## 5     21</code></pre>
<p>The matrices are in RDS format. For learning it is probably best to process them to remove any reads containing NAs.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>i=<span class="dv">1</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">dim</span>(dataMatrix)</span></code></pre></div>
<pre><code>## [1] 1690   25</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>dataMatrix&lt;-<span class="kw">removeAllNArows</span>(dataMatrix)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw">dim</span>(dataMatrix)</span></code></pre></div>
<pre><code>## [1] 1068   25</code></pre>
<div id="learning-classes-for-single-genes" class="section level2">
<h2>Learning classes for single genes</h2>
<p>To look at one gene at a time one can use the minimal matrix which includes only positions with a CpG or GpC motif in them.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>k_range =<span class="st"> </span><span class="dv">2</span><span class="op">:</span><span class="dv">8</span>      <span class="co"># Number of classes to be found</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>maxIterations =<span class="st"> </span><span class="dv">100</span> <span class="co"># number of iterations of EM clustering to perform if it does not converge</span></span>
<span id="cb7-3"><a href="#cb7-3"></a>convergenceError =<span class="st"> </span><span class="fl">10e-6</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>numRepeats=<span class="dv">10</span> <span class="co"># number of repeats of clustering each matrix (to account for fraction of methylation)</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>xRange=<span class="kw">c</span>(<span class="op">-</span><span class="dv">250</span>,<span class="dv">250</span>)</span>
<span id="cb7-6"><a href="#cb7-6"></a>maxB=<span class="dv">50</span> <span class="co"># Number of randomised matrices to generate</span></span>
<span id="cb7-7"><a href="#cb7-7"></a>outPath=<span class="st">&quot;./EMres&quot;</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>maxTasks=<span class="dv">4</span></span>
<span id="cb7-9"><a href="#cb7-9"></a>taskId=<span class="dv">1</span></span>
<span id="cb7-10"><a href="#cb7-10"></a>nThreads=<span class="dv">4</span></span>
<span id="cb7-11"><a href="#cb7-11"></a>setSeed=<span class="ot">FALSE</span></span>
<span id="cb7-12"><a href="#cb7-12"></a>distMetric=<span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;cosineDist&quot;</span>)</span>
<span id="cb7-13"><a href="#cb7-13"></a></span>
<span id="cb7-14"><a href="#cb7-14"></a><span class="cf">if</span>(<span class="op">!</span><span class="kw">dir.exists</span>(outPath)){</span>
<span id="cb7-15"><a href="#cb7-15"></a>  <span class="kw">dir.create</span>(outPath)</span>
<span id="cb7-16"><a href="#cb7-16"></a>}</span>
<span id="cb7-17"><a href="#cb7-17"></a></span>
<span id="cb7-18"><a href="#cb7-18"></a><span class="co">#split table indicies into nTasks number of groups</span></span>
<span id="cb7-19"><a href="#cb7-19"></a>taskSubList&lt;-<span class="kw">split</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(matTable),<span class="kw">sort</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(matTable)<span class="op">%%</span>maxTasks))</span>
<span id="cb7-20"><a href="#cb7-20"></a></span>
<span id="cb7-21"><a href="#cb7-21"></a><span class="kw">set.seed</span>(<span class="dv">200413</span>)</span>
<span id="cb7-22"><a href="#cb7-22"></a><span class="cf">for</span> (i <span class="cf">in</span> taskSubList[[taskId]]){</span>
<span id="cb7-23"><a href="#cb7-23"></a><span class="co">#for (i in 1:nrow(matTable)) {</span></span>
<span id="cb7-24"><a href="#cb7-24"></a>  regionName=matTable<span class="op">$</span>region[i]</span>
<span id="cb7-25"><a href="#cb7-25"></a>  sampleName=matTable<span class="op">$</span>sample[i]</span>
<span id="cb7-26"><a href="#cb7-26"></a>  outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb7-27"><a href="#cb7-27"></a>  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Clustering&quot;</span>, outFileBase))</span>
<span id="cb7-28"><a href="#cb7-28"></a>  dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], </span>
<span id="cb7-29"><a href="#cb7-29"></a>                              <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</span>
<span id="cb7-30"><a href="#cb7-30"></a>  <span class="kw">dim</span>(dataMatrix)</span>
<span id="cb7-31"><a href="#cb7-31"></a>  dataMatrix&lt;-<span class="kw">removeAllNArows</span>(dataMatrix)</span>
<span id="cb7-32"><a href="#cb7-32"></a>  <span class="kw">dim</span>(dataMatrix)</span>
<span id="cb7-33"><a href="#cb7-33"></a>  </span>
<span id="cb7-34"><a href="#cb7-34"></a>  allClassMeans&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb7-35"><a href="#cb7-35"></a>    {</span>
<span id="cb7-36"><a href="#cb7-36"></a>      <span class="kw">print</span>(<span class="st">&quot;running EM for a range of class numbers&quot;</span>)</span>
<span id="cb7-37"><a href="#cb7-37"></a>      <span class="kw">runEMrangeClassNum</span>(dataMatrix, k_range, convergenceError, maxIterations,</span>
<span id="cb7-38"><a href="#cb7-38"></a>                     <span class="dt">EMrepeats=</span>numRepeats, <span class="dt">outPath=</span>outPath, <span class="dt">xRange=</span>xRange,</span>
<span id="cb7-39"><a href="#cb7-39"></a>                     <span class="dt">outFileBase=</span><span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>),</span>
<span id="cb7-40"><a href="#cb7-40"></a>                     <span class="dt">doIndividualPlots=</span><span class="ot">FALSE</span>, <span class="dt">distMetric=</span>distMetric)</span>
<span id="cb7-41"><a href="#cb7-41"></a></span>
<span id="cb7-42"><a href="#cb7-42"></a>    },</span>
<span id="cb7-43"><a href="#cb7-43"></a>      <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb7-44"><a href="#cb7-44"></a>  )</span>
<span id="cb7-45"><a href="#cb7-45"></a></span>
<span id="cb7-46"><a href="#cb7-46"></a>  <span class="cf">if</span>(<span class="kw">is.list</span>(allClassMeans)){</span>
<span id="cb7-47"><a href="#cb7-47"></a>        <span class="kw">saveRDS</span>(allClassMeans,<span class="kw">paste0</span>(outPath,<span class="st">&quot;/allClassMeans_&quot;</span>,outFileBase,<span class="st">&quot;.rds&quot;</span>))</span>
<span id="cb7-48"><a href="#cb7-48"></a>  } <span class="cf">else</span> {</span>
<span id="cb7-49"><a href="#cb7-49"></a>        <span class="kw">print</span>(allClassMeans) <span class="co"># error message</span></span>
<span id="cb7-50"><a href="#cb7-50"></a>  }</span>
<span id="cb7-51"><a href="#cb7-51"></a></span>
<span id="cb7-52"><a href="#cb7-52"></a>  clustMetrics&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb7-53"><a href="#cb7-53"></a>    {</span>
<span id="cb7-54"><a href="#cb7-54"></a>        <span class="kw">print</span>(<span class="st">&quot;plotting clustering metrics for a range of class sizes&quot;</span>)</span>
<span id="cb7-55"><a href="#cb7-55"></a>        <span class="kw">plotClusteringMetrics</span>(dataMatrix, k_range, maxB, convergenceError,</span>
<span id="cb7-56"><a href="#cb7-56"></a>            maxIterations, outPath, outFileBase, <span class="dt">EMrep=</span><span class="ot">NULL</span>, <span class="dt">nThreads=</span>nThreads,</span>
<span id="cb7-57"><a href="#cb7-57"></a>            <span class="dt">setSeed=</span>setSeed, <span class="dt">distMetric=</span>distMetric)</span>
<span id="cb7-58"><a href="#cb7-58"></a>    },</span>
<span id="cb7-59"><a href="#cb7-59"></a>    <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb7-60"><a href="#cb7-60"></a>  )</span>
<span id="cb7-61"><a href="#cb7-61"></a>  <span class="cf">if</span>(<span class="kw">length</span>(clustMetrics)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb7-62"><a href="#cb7-62"></a>    <span class="kw">print</span>(clustMetrics)</span>
<span id="cb7-63"><a href="#cb7-63"></a>  }</span>
<span id="cb7-64"><a href="#cb7-64"></a></span>
<span id="cb7-65"><a href="#cb7-65"></a>  pcaPlots&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb7-66"><a href="#cb7-66"></a>    {</span>
<span id="cb7-67"><a href="#cb7-67"></a>      <span class="kw">print</span>(<span class="st">&quot;plotting PCA of clusters&quot;</span>)</span>
<span id="cb7-68"><a href="#cb7-68"></a>      <span class="kw">plotPCAofMatrixClasses</span>(k_range, outPath, outFileBase)</span>
<span id="cb7-69"><a href="#cb7-69"></a>    },</span>
<span id="cb7-70"><a href="#cb7-70"></a>    <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb7-71"><a href="#cb7-71"></a>   )</span>
<span id="cb7-72"><a href="#cb7-72"></a>  <span class="cf">if</span>(<span class="kw">length</span>(pcaPlots)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb7-73"><a href="#cb7-73"></a>    <span class="kw">print</span>(pcaPlots)</span>
<span id="cb7-74"><a href="#cb7-74"></a>  }</span>
<span id="cb7-75"><a href="#cb7-75"></a></span>
<span id="cb7-76"><a href="#cb7-76"></a>  umapPlots&lt;-<span class="kw">tryCatch</span>( </span>
<span id="cb7-77"><a href="#cb7-77"></a>    {</span>
<span id="cb7-78"><a href="#cb7-78"></a>      <span class="kw">print</span>(<span class="st">&quot;plotting UMAP of clusters&quot;</span>)</span>
<span id="cb7-79"><a href="#cb7-79"></a>      <span class="kw">plotUMAPofMatrixClasses</span>(k_range, outPath, outFileBase)</span>
<span id="cb7-80"><a href="#cb7-80"></a>    },</span>
<span id="cb7-81"><a href="#cb7-81"></a>    <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb7-82"><a href="#cb7-82"></a>   )</span>
<span id="cb7-83"><a href="#cb7-83"></a>  <span class="cf">if</span>(<span class="kw">length</span>(umapPlots)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb7-84"><a href="#cb7-84"></a>    <span class="kw">print</span>(umapPlots)</span>
<span id="cb7-85"><a href="#cb7-85"></a>  }</span>
<span id="cb7-86"><a href="#cb7-86"></a>  </span>
<span id="cb7-87"><a href="#cb7-87"></a>}</span></code></pre></div>
</div>
<div id="plotting-classes-with-track-data" class="section level2">
<h2>Plotting classes with track data</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>i=<span class="dv">1</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</span>
<span id="cb8-3"><a href="#cb8-3"></a>regionName=matTable<span class="op">$</span>region[i]</span>
<span id="cb8-4"><a href="#cb8-4"></a>sampleName=matTable<span class="op">$</span>sample[i]</span>
<span id="cb8-5"><a href="#cb8-5"></a>outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb8-6"><a href="#cb8-6"></a></span>
<span id="cb8-7"><a href="#cb8-7"></a>allClassMeansList&lt;-<span class="kw">readRDS</span>(<span class="st">&quot;/Users/semple/Documents/MeisterLab/myPackages/EMclassifieR/inst/extdata/rds/allClassMeans_dS02-182_WBGene00015947.rds&quot;</span>)</span>
<span id="cb8-8"><a href="#cb8-8"></a>numClasses=<span class="dv">5</span></span>
<span id="cb8-9"><a href="#cb8-9"></a>allClassMeans&lt;-allClassMeansList[[numClasses]]</span>
<span id="cb8-10"><a href="#cb8-10"></a>allClassMeans</span>
<span id="cb8-11"><a href="#cb8-11"></a></span>
<span id="cb8-12"><a href="#cb8-12"></a>ampTSS&lt;-<span class="kw">readRDS</span>(<span class="st">&#39;/Users/semple/Documents/MeisterLab/myPackages/EMclassifieR/inst/extdata/rds/ampliconMaxTSSgr.RDS&#39;</span>)</span>
<span id="cb8-13"><a href="#cb8-13"></a></span>
<span id="cb8-14"><a href="#cb8-14"></a></span>
<span id="cb8-15"><a href="#cb8-15"></a>winSize=<span class="dv">500</span></span>
<span id="cb8-16"><a href="#cb8-16"></a>tssWin&lt;-ampTSS</span>
<span id="cb8-17"><a href="#cb8-17"></a>GenomicRanges<span class="op">::</span><span class="kw">mcols</span>(tssWin)<span class="op">$</span>TSS&lt;-GenomicRanges<span class="op">::</span><span class="kw">start</span>(tssWin)</span>
<span id="cb8-18"><a href="#cb8-18"></a>tssWin&lt;-GenomicRanges<span class="op">::</span><span class="kw">resize</span>(tssWin,<span class="dt">width=</span>winSize,<span class="dt">fix=</span><span class="st">&quot;center&quot;</span>)</span>
<span id="cb8-19"><a href="#cb8-19"></a></span>
<span id="cb8-20"><a href="#cb8-20"></a></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="kw">names</span>(GenomicRanges<span class="op">::</span><span class="kw">mcols</span>(tssWin))[<span class="dv">1</span>]&lt;-<span class="st">&quot;ID&quot;</span></span>
<span id="cb8-22"><a href="#cb8-22"></a>regionGR&lt;-tssWin[i]</span>
<span id="cb8-23"><a href="#cb8-23"></a>featureGR&lt;-ampTSS[i]</span>
<span id="cb8-24"><a href="#cb8-24"></a>regionName=matTable<span class="op">$</span>region[i]</span>
<span id="cb8-25"><a href="#cb8-25"></a>sampleName=matTable<span class="op">$</span>sample[i]</span>
<span id="cb8-26"><a href="#cb8-26"></a>outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb8-27"><a href="#cb8-27"></a><span class="co">#seqlevels(ampTSS)&lt;-seqlevels(Celegans)</span></span>
<span id="cb8-28"><a href="#cb8-28"></a><span class="co">#seqinfo(ampTSS)&lt;-seqinfo(Celegans)</span></span>
<span id="cb8-29"><a href="#cb8-29"></a><span class="co">#saveRDS(ampTSS,&#39;/Users/semple/Documents/MeisterLab/myPackages/EMclassifieR/inst/extdata/rds/ampliconMaxTSSgr.RDS&#39;)</span></span>
<span id="cb8-30"><a href="#cb8-30"></a></span>
<span id="cb8-31"><a href="#cb8-31"></a><span class="co"># genomeVer=&quot;WS275&quot;</span></span>
<span id="cb8-32"><a href="#cb8-32"></a><span class="co"># genomeDir=paste0(&quot;~/Documents/MeisterLab/GenomeVer/&quot;,genomeVer)</span></span>
<span id="cb8-33"><a href="#cb8-33"></a><span class="co"># txdb&lt;-AnnotationDbi::loadDb(paste0(genomeDir, </span></span>
<span id="cb8-34"><a href="#cb8-34"></a><span class="co">#                                    &quot;/annotations/c_elegans.PRJNA13758.&quot;,</span></span>
<span id="cb8-35"><a href="#cb8-35"></a><span class="co">#                                    genomeVer,  &quot;.annotations.sqlite&quot;))</span></span>
<span id="cb8-36"><a href="#cb8-36"></a><span class="co"># seqlevelsStyle(txdb)&lt;-&quot;ucsc&quot;</span></span>
<span id="cb8-37"><a href="#cb8-37"></a><span class="co"># GenomeInfoDb::genome(txdb)&lt;-GenomeInfoDb::genome(regionGR)</span></span>
<span id="cb8-38"><a href="#cb8-38"></a><span class="co">#library(&quot;TxDb.Celegans.UCSC.ce11.refGene&quot;)</span></span>
<span id="cb8-39"><a href="#cb8-39"></a>txdb &lt;-<span class="st"> </span>TxDb.Celegans.UCSC.ce11.refGene<span class="op">::</span>TxDb.Celegans.UCSC.ce11.refGene</span>
<span id="cb8-40"><a href="#cb8-40"></a>bigwigListFile&lt;-<span class="st">&quot;~/Documents/MeisterLab/sequencingData/public/bigwigFileList.txt&quot;</span></span>
<span id="cb8-41"><a href="#cb8-41"></a><span class="co">#bigwigListFile&lt;-&quot;~/Documents/MeisterLab/sequencingData/public/bigwigFileList_kranz.txt&quot;</span></span>
<span id="cb8-42"><a href="#cb8-42"></a>strandedBwFile&lt;-<span class="st">&quot;~/Documents/MeisterLab/sequencingData/public/strandedBwFile.txt&quot;</span></span>
<span id="cb8-43"><a href="#cb8-43"></a>gr&lt;-<span class="kw">allClassMeansToGR</span>(allClassMeans, regionGR)</span>
<span id="cb8-44"><a href="#cb8-44"></a></span>
<span id="cb8-45"><a href="#cb8-45"></a>dTrack&lt;-Gviz<span class="op">::</span><span class="kw">DataTrack</span>(gr,<span class="dt">name=</span><span class="st">&quot;dSMF Classes&quot;</span>)</span>
<span id="cb8-46"><a href="#cb8-46"></a>Gviz<span class="op">::</span><span class="kw">plotTracks</span>(dTrack, <span class="dt">groups=</span><span class="kw">rep</span>(<span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">&quot;class&quot;</span>,<span class="dv">1</span><span class="op">:</span>numClasses)), <span class="dt">times=</span><span class="dv">10</span>),</span>
<span id="cb8-47"><a href="#cb8-47"></a>                 <span class="dt">type=</span><span class="kw">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;p&quot;</span>,<span class="st">&quot;confint&quot;</span>))</span>
<span id="cb8-48"><a href="#cb8-48"></a></span>
<span id="cb8-49"><a href="#cb8-49"></a><span class="kw">print</span>(txdb)</span>
<span id="cb8-50"><a href="#cb8-50"></a>outPath=<span class="st">&quot;./EMres&quot;</span></span>
<span id="cb8-51"><a href="#cb8-51"></a>grDevices<span class="op">::</span><span class="kw">pdf</span>(<span class="kw">paste0</span>(outPath,<span class="st">&quot;/tracks_&quot;</span>,</span>
<span id="cb8-52"><a href="#cb8-52"></a>                          outFileBase,<span class="st">&quot;_K&quot;</span>,</span>
<span id="cb8-53"><a href="#cb8-53"></a>                          numClasses, <span class="st">&quot;.pdf&quot;</span>),</span>
<span id="cb8-54"><a href="#cb8-54"></a>                   <span class="dt">paper=</span><span class="st">&quot;a4&quot;</span>, <span class="dt">height=</span><span class="dv">11</span>, <span class="dt">width=</span><span class="dv">8</span>)</span>
<span id="cb8-55"><a href="#cb8-55"></a></span>
<span id="cb8-56"><a href="#cb8-56"></a><span class="kw">plotClassMeansWithTracks</span>(allClassMeans, regionGR, <span class="st">&quot;middle&quot;</span>, txdb, featureGR, </span>
<span id="cb8-57"><a href="#cb8-57"></a>                         bigwigListFile, strandedBwFile)</span>
<span id="cb8-58"><a href="#cb8-58"></a><span class="kw">dev.off</span>()</span></code></pre></div>
</div>
<div id="multigene-classification" class="section level2">
<h2>Multigene classification</h2>
<p>First you must create a data set combining and equal amount of reads from different matrices</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>tablePath=<span class="st">&quot;csv/MatrixLog_relCoord_ampTSS.csv&quot;</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>matTable&lt;-<span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, tablePath, <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3"></a>                               <span class="dt">mustWork=</span><span class="ot">TRUE</span>), <span class="dt">stringsAsFactors=</span>F)</span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="kw">head</span>(matTable)</span>
<span id="cb9-5"><a href="#cb9-5"></a>multiGeneMat&lt;-<span class="ot">NULL</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(matTable[matTable<span class="op">$</span>sample<span class="op">==</span><span class="st">&quot;dS02-182&quot;</span>])){</span>
<span id="cb9-7"><a href="#cb9-7"></a>  regionName=matTable<span class="op">$</span>region[i]</span>
<span id="cb9-8"><a href="#cb9-8"></a>  sampleName=matTable<span class="op">$</span>sample[i]</span>
<span id="cb9-9"><a href="#cb9-9"></a>  outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb9-10"><a href="#cb9-10"></a>  dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], </span>
<span id="cb9-11"><a href="#cb9-11"></a>                                  <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</span>
<span id="cb9-12"><a href="#cb9-12"></a>  </span>
<span id="cb9-13"><a href="#cb9-13"></a>  subMatrix&lt;-<span class="kw">selectReadsFromMatrix</span>(dataMatrix,<span class="dt">minReads=</span><span class="dv">50</span>,</span>
<span id="cb9-14"><a href="#cb9-14"></a>                                 <span class="dt">addToReadName=</span>outFileBase,</span>
<span id="cb9-15"><a href="#cb9-15"></a>                                 <span class="dt">preferBest=</span>T)</span>
<span id="cb9-16"><a href="#cb9-16"></a>  <span class="cf">if</span>(<span class="op">!</span><span class="kw">is.null</span>(subMatrix)){</span>
<span id="cb9-17"><a href="#cb9-17"></a>    fullMatrix&lt;-<span class="kw">getFullMatrix</span>(subMatrix)</span>
<span id="cb9-18"><a href="#cb9-18"></a>    winMatrix&lt;-<span class="kw">prepareWindows</span>(fullMatrix)</span>
<span id="cb9-19"><a href="#cb9-19"></a>  </span>
<span id="cb9-20"><a href="#cb9-20"></a>    <span class="cf">if</span>(<span class="kw">is.null</span>(multiGeneMat)){</span>
<span id="cb9-21"><a href="#cb9-21"></a>      multiGeneMat&lt;-winMatrix</span>
<span id="cb9-22"><a href="#cb9-22"></a>    } <span class="cf">else</span> {</span>
<span id="cb9-23"><a href="#cb9-23"></a>      multiGeneMat&lt;-<span class="kw">rbind</span>(multiGeneMat,winMatrix)</span>
<span id="cb9-24"><a href="#cb9-24"></a>    }</span>
<span id="cb9-25"><a href="#cb9-25"></a>  }</span>
<span id="cb9-26"><a href="#cb9-26"></a>}</span>
<span id="cb9-27"><a href="#cb9-27"></a></span>
<span id="cb9-28"><a href="#cb9-28"></a>multiGeneMat&lt;-<span class="kw">rescale_minus1To1</span>(multiGeneMat)</span>
<span id="cb9-29"><a href="#cb9-29"></a>multiGeneMat&lt;-<span class="kw">rescale_0To1</span>(multiGeneMat)</span>
<span id="cb9-30"><a href="#cb9-30"></a><span class="co">#multiGeneMat1&lt;-recodeMatrixAsNumeric(multiGeneMat)</span></span>
<span id="cb9-31"><a href="#cb9-31"></a></span>
<span id="cb9-32"><a href="#cb9-32"></a>k_range =<span class="st"> </span><span class="dv">2</span><span class="op">:</span><span class="dv">8</span>      <span class="co"># Number of classes to be found</span></span>
<span id="cb9-33"><a href="#cb9-33"></a>maxIterations =<span class="st"> </span><span class="dv">100</span> <span class="co"># number of iterations of EM clustering to perform if it does not converge</span></span>
<span id="cb9-34"><a href="#cb9-34"></a>convergenceError =<span class="st"> </span><span class="fl">10e-6</span></span>
<span id="cb9-35"><a href="#cb9-35"></a>numRepeats=<span class="dv">10</span> <span class="co"># number of repeats of clustering each matrix (to account for fraction of methylation)</span></span>
<span id="cb9-36"><a href="#cb9-36"></a>xRange=<span class="kw">c</span>(<span class="op">-</span><span class="dv">250</span>,<span class="dv">250</span>)</span>
<span id="cb9-37"><a href="#cb9-37"></a>maxB=<span class="dv">50</span> <span class="co"># Number of randomised matrices to generate</span></span>
<span id="cb9-38"><a href="#cb9-38"></a>outPath=<span class="st">&quot;./EMres_multigene_cosine&quot;</span></span>
<span id="cb9-39"><a href="#cb9-39"></a>maxTasks=<span class="dv">4</span></span>
<span id="cb9-40"><a href="#cb9-40"></a>taskId=<span class="dv">1</span></span>
<span id="cb9-41"><a href="#cb9-41"></a>nThreads=<span class="dv">4</span></span>
<span id="cb9-42"><a href="#cb9-42"></a>setSeed=<span class="ot">FALSE</span></span>
<span id="cb9-43"><a href="#cb9-43"></a>distMetric=<span class="kw">list</span>(<span class="dt">name=</span><span class="st">&quot;cosineDist&quot;</span>,<span class="dt">rescale=</span>F)</span>
<span id="cb9-44"><a href="#cb9-44"></a></span>
<span id="cb9-45"><a href="#cb9-45"></a><span class="cf">if</span>(<span class="op">!</span><span class="kw">dir.exists</span>(outPath)){</span>
<span id="cb9-46"><a href="#cb9-46"></a>  <span class="kw">dir.create</span>(outPath)</span>
<span id="cb9-47"><a href="#cb9-47"></a>}</span>
<span id="cb9-48"><a href="#cb9-48"></a></span>
<span id="cb9-49"><a href="#cb9-49"></a></span>
<span id="cb9-50"><a href="#cb9-50"></a><span class="kw">set.seed</span>(<span class="dv">200413</span>)</span>
<span id="cb9-51"><a href="#cb9-51"></a></span>
<span id="cb9-52"><a href="#cb9-52"></a>regionName=<span class="st">&quot;multiGene&quot;</span></span>
<span id="cb9-53"><a href="#cb9-53"></a>sampleName=<span class="st">&quot;dS02-182&quot;</span></span>
<span id="cb9-54"><a href="#cb9-54"></a>outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</span>
<span id="cb9-55"><a href="#cb9-55"></a><span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Clustering&quot;</span>, outFileBase))</span>
<span id="cb9-56"><a href="#cb9-56"></a>dataMatrix&lt;-multiGeneMat</span>
<span id="cb9-57"><a href="#cb9-57"></a><span class="kw">dim</span>(dataMatrix)</span>
<span id="cb9-58"><a href="#cb9-58"></a>dataMatrix&lt;-<span class="kw">removeAllNArows</span>(dataMatrix)</span>
<span id="cb9-59"><a href="#cb9-59"></a><span class="kw">dim</span>(dataMatrix)</span>
<span id="cb9-60"><a href="#cb9-60"></a></span>
<span id="cb9-61"><a href="#cb9-61"></a>allClassMeans&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb9-62"><a href="#cb9-62"></a>  {</span>
<span id="cb9-63"><a href="#cb9-63"></a>    <span class="kw">print</span>(<span class="st">&quot;running EM for a range of class numbers&quot;</span>)</span>
<span id="cb9-64"><a href="#cb9-64"></a>    <span class="kw">runEMrangeClassNum</span>(dataMatrix, k_range, convergenceError, maxIterations,</span>
<span id="cb9-65"><a href="#cb9-65"></a>                       <span class="dt">EMrepeats=</span>numRepeats, <span class="dt">outPath=</span>outPath, <span class="dt">xRange=</span>xRange,</span>
<span id="cb9-66"><a href="#cb9-66"></a>                       <span class="dt">outFileBase=</span><span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>),</span>
<span id="cb9-67"><a href="#cb9-67"></a>                       <span class="dt">doIndividualPlots=</span><span class="ot">FALSE</span>, <span class="dt">distMetric=</span>distMetric)</span>
<span id="cb9-68"><a href="#cb9-68"></a>    </span>
<span id="cb9-69"><a href="#cb9-69"></a>  },</span>
<span id="cb9-70"><a href="#cb9-70"></a>  <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb9-71"><a href="#cb9-71"></a>)</span>
<span id="cb9-72"><a href="#cb9-72"></a></span>
<span id="cb9-73"><a href="#cb9-73"></a><span class="cf">if</span>(<span class="kw">is.list</span>(allClassMeans)){</span>
<span id="cb9-74"><a href="#cb9-74"></a>  <span class="kw">saveRDS</span>(allClassMeans,<span class="kw">paste0</span>(outPath,<span class="st">&quot;/allClassMeans_&quot;</span>,outFileBase,<span class="st">&quot;.rds&quot;</span>))</span>
<span id="cb9-75"><a href="#cb9-75"></a>} <span class="cf">else</span> {</span>
<span id="cb9-76"><a href="#cb9-76"></a>  <span class="kw">print</span>(allClassMeans) <span class="co"># error message</span></span>
<span id="cb9-77"><a href="#cb9-77"></a>}</span>
<span id="cb9-78"><a href="#cb9-78"></a></span>
<span id="cb9-79"><a href="#cb9-79"></a>clustMetrics&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb9-80"><a href="#cb9-80"></a>  {</span>
<span id="cb9-81"><a href="#cb9-81"></a>    <span class="kw">print</span>(<span class="st">&quot;plotting clustering metrics for a range of class sizes&quot;</span>)</span>
<span id="cb9-82"><a href="#cb9-82"></a>    <span class="kw">plotClusteringMetrics</span>(dataMatrix, k_range, maxB, convergenceError,</span>
<span id="cb9-83"><a href="#cb9-83"></a>                          maxIterations, outPath, outFileBase, <span class="dt">EMrep=</span><span class="ot">NULL</span>,</span>
<span id="cb9-84"><a href="#cb9-84"></a>                          <span class="dt">nThreads=</span>nThreads,</span>
<span id="cb9-85"><a href="#cb9-85"></a>                          <span class="dt">setSeed=</span>setSeed, <span class="dt">distMetric=</span>distMetric)</span>
<span id="cb9-86"><a href="#cb9-86"></a>  },</span>
<span id="cb9-87"><a href="#cb9-87"></a>  <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb9-88"><a href="#cb9-88"></a>)</span>
<span id="cb9-89"><a href="#cb9-89"></a><span class="cf">if</span>(<span class="kw">length</span>(clustMetrics)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb9-90"><a href="#cb9-90"></a>  <span class="kw">print</span>(clustMetrics)</span>
<span id="cb9-91"><a href="#cb9-91"></a>}</span>
<span id="cb9-92"><a href="#cb9-92"></a></span>
<span id="cb9-93"><a href="#cb9-93"></a>pcaPlots&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb9-94"><a href="#cb9-94"></a>  {</span>
<span id="cb9-95"><a href="#cb9-95"></a>    <span class="kw">print</span>(<span class="st">&quot;plotting PCA of clusters&quot;</span>)</span>
<span id="cb9-96"><a href="#cb9-96"></a>    <span class="kw">plotPCAofMatrixClasses</span>(k_range, outPath, outFileBase)</span>
<span id="cb9-97"><a href="#cb9-97"></a>  },</span>
<span id="cb9-98"><a href="#cb9-98"></a>  <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb9-99"><a href="#cb9-99"></a>)</span>
<span id="cb9-100"><a href="#cb9-100"></a><span class="cf">if</span>(<span class="kw">length</span>(pcaPlots)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb9-101"><a href="#cb9-101"></a>  <span class="kw">print</span>(pcaPlots)</span>
<span id="cb9-102"><a href="#cb9-102"></a>}</span>
<span id="cb9-103"><a href="#cb9-103"></a></span>
<span id="cb9-104"><a href="#cb9-104"></a>umapPlots&lt;-<span class="kw">tryCatch</span>(</span>
<span id="cb9-105"><a href="#cb9-105"></a>  {</span>
<span id="cb9-106"><a href="#cb9-106"></a>    <span class="kw">print</span>(<span class="st">&quot;plotting UMAP of clusters&quot;</span>)</span>
<span id="cb9-107"><a href="#cb9-107"></a>    <span class="kw">plotUMAPofMatrixClasses</span>(k_range, outPath, outFileBase)</span>
<span id="cb9-108"><a href="#cb9-108"></a>  },</span>
<span id="cb9-109"><a href="#cb9-109"></a>  <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</span>
<span id="cb9-110"><a href="#cb9-110"></a>)</span>
<span id="cb9-111"><a href="#cb9-111"></a><span class="cf">if</span>(<span class="kw">length</span>(umapPlots)<span class="op">==</span><span class="dv">1</span>) {</span>
<span id="cb9-112"><a href="#cb9-112"></a>  <span class="kw">print</span>(umapPlots)</span>
<span id="cb9-113"><a href="#cb9-113"></a>}</span>
<span id="cb9-114"><a href="#cb9-114"></a></span>
<span id="cb9-115"><a href="#cb9-115"></a></span>
<span id="cb9-116"><a href="#cb9-116"></a><span class="co"># to use -1 to 1</span></span>
<span id="cb9-117"><a href="#cb9-117"></a><span class="co">#https://stats.stackexchange.com/questions/256917/can-an-unstandarized-beta-distribution-have-a-negative-domain</span></span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
