---
title: "Using EMclassifieR on methylation data"
author: "Jennifer Semple"
date: "7/18/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(EMclassifieR, quietly=T)
```

# The data
DSMF data comes as matrices of reads x positions where each row refers to a single (usually paired-end) read and each column refers to a CpG or GpC position where methylation has been scored. The positions should be in relative coordinates (e.g. -250 to +250 around the TSS), so that matrices from different genes can be compared. 
Some example data is included with the package. First we read in a table with a list of the matrix filenames (including path). The important columns in this table are "filename" and "reads".
```{r}
tablePath="csv/MatrixLog_relCoord_ampTSS.csv"
matTable<-read.csv(system.file("extdata", tablePath, package = "EMclassifieR", mustWork=TRUE),stringsAsFactors=F)
head(matTable)
```

The matrices are in RDS format. For learning it is probably best to process them to remove any reads containing NAs.
```{r}
m1<-readRDS(system.file("extdata", matTable$filename[1], package = "EMclassifieR", mustWork=TRUE))
dim(m1)
m1<-removeAllNArows(m1)
dim(m1)
```


## Learning classes for single genes
To look at one gene at a time one can use the minimal matrix which includes only positions with a CpG or GpC motif in them.

```{r}
# Configuration
K_range = 2:10                       # Number of classes to be found
maxIterations = 100
convergence_error = 10^-6
#geneNr = 1
performSilhouette = T
silhouetteDistance = "euclidean"

geneNr=1

gene = matTable$region[geneNr]
# Matrix containing the samples. data[i,j] is the bincount of sample i at position j.
data = m1

# orders the clusters in the plot according to the first plot for easier visual comparison
orderByPreviousClusters = F
if (exists("firstClusteringClassMeans") && !is.null(firstClusteringClassMeans)) {
  orderByPreviousClusters = T
} else {
  firstClusteringClassMeans = vector(mode="list", length=length(K_range))
}

# Adjusted from: http://www.sthda.com/english/wiki/print.php?id=239#average-silhouette-method
# vector of Silhouette values for K determination
if (performSilhouette) {
  sil <- rep(0, max(K_range))
  dis = dist(data, method = silhouetteDistance) # might not work for >2000 reads (huge matrix)
}



includeInFileName=numClasses
  ggplot2::ggsave(paste0(workDir,"/plots/singleMoleculePlots_",regionType,"/",gene,
                         "_", includeInFileName,".png"),
                  plot=p, device="png", width=29, height=20, units="cm")
```

