<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jennifer Semple" />

<meta name="date" content="2020-04-17" />

<title>Using EMclassifieR on methylation data</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Using EMclassifieR on methylation data</h1>
<h4 class="author">Jennifer Semple</h4>
<h4 class="date">2020-04-17</h4>



<div id="the-data" class="section level1">
<h1>The data</h1>
<p>DSMF data comes as matrices of reads x positions where each row refers to a single (usually paired-end) read and each column refers to a CpG or GpC position where methylation has been scored. The positions should be in relative coordinates (e.g. -250 to +250 around the TSS), so that matrices from different genes can be compared. Some example data is included with the package. First we read in a table with a list of the matrix filenames (including path). The important columns in this table are “filename” and “reads”.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">tablePath=<span class="st">&quot;csv/MatrixLog_relCoord_ampTSS.csv&quot;</span></a>
<a class="sourceLine" id="cb1-2" title="2">matTable&lt;-<span class="kw">read.csv</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, tablePath, <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>,</a>
<a class="sourceLine" id="cb1-3" title="3">                               <span class="dt">mustWork=</span><span class="ot">TRUE</span>), <span class="dt">stringsAsFactors=</span>F)</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw">head</span>(matTable)</a></code></pre></div>
<pre><code>##                                          filename   sample         region reads
## 1 rds/relCoord_ampTSS/dS02-182_WBGene00015947.rds dS02-182 WBGene00015947  1690
## 2 rds/relCoord_ampTSS/dS02-182_WBGene00015955.rds dS02-182 WBGene00015955 10677
## 3 rds/relCoord_ampTSS/dS02-182_WBGene00009620.rds dS02-182 WBGene00009620  3416
## 4 rds/relCoord_ampTSS/dS02-182_WBGene00009621.rds dS02-182 WBGene00009621    31
## 5 rds/relCoord_ampTSS/dS02-182_WBGene00009801.rds dS02-182 WBGene00009801  1177
##   motifs
## 1     25
## 2     36
## 3     26
## 4     25
## 5     21</code></pre>
<p>The matrices are in RDS format. For learning it is probably best to process them to remove any reads containing NAs.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">i=<span class="dv">1</span></a>
<a class="sourceLine" id="cb3-2" title="2">dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="kw">dim</span>(dataMatrix)</a></code></pre></div>
<pre><code>## [1] 1690   25</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">dataMatrix&lt;-<span class="kw">removeAllNArows</span>(dataMatrix)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw">dim</span>(dataMatrix)</a></code></pre></div>
<pre><code>## [1] 1068   25</code></pre>
<div id="learning-classes-for-single-genes" class="section level2">
<h2>Learning classes for single genes</h2>
<p>To look at one gene at a time one can use the minimal matrix which includes only positions with a CpG or GpC motif in them.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">k_range =<span class="st"> </span><span class="dv">2</span><span class="op">:</span><span class="dv">8</span>      <span class="co"># Number of classes to be found</span></a>
<a class="sourceLine" id="cb7-2" title="2">maxIterations =<span class="st"> </span><span class="dv">100</span> <span class="co"># number of iterations of EM clustering to perform if it does not converge</span></a>
<a class="sourceLine" id="cb7-3" title="3">convergenceError =<span class="st"> </span><span class="fl">10e-6</span></a>
<a class="sourceLine" id="cb7-4" title="4">numRepeats=<span class="dv">10</span> <span class="co"># number of repeats of clustering each matrix (to account for fraction of methylation)</span></a>
<a class="sourceLine" id="cb7-5" title="5">xRange=<span class="kw">c</span>(<span class="op">-</span><span class="dv">250</span>,<span class="dv">250</span>)</a>
<a class="sourceLine" id="cb7-6" title="6">maxB=<span class="dv">100</span> <span class="co"># Number of randomised matrices to generate</span></a>
<a class="sourceLine" id="cb7-7" title="7">outPath=<span class="st">&quot;./EMres&quot;</span></a>
<a class="sourceLine" id="cb7-8" title="8">maxTasks=<span class="dv">4</span></a>
<a class="sourceLine" id="cb7-9" title="9">taskId=<span class="dv">1</span></a>
<a class="sourceLine" id="cb7-10" title="10">nThreads=<span class="dv">1</span></a>
<a class="sourceLine" id="cb7-11" title="11">setSeed=<span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb7-12" title="12"></a>
<a class="sourceLine" id="cb7-13" title="13"><span class="co">#split table indecies into nTasks number of groups</span></a>
<a class="sourceLine" id="cb7-14" title="14">taskSubList&lt;-<span class="kw">split</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(matTable),<span class="kw">sort</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(matTable)<span class="op">%%</span>maxTasks))</a>
<a class="sourceLine" id="cb7-15" title="15"></a>
<a class="sourceLine" id="cb7-16" title="16"><span class="kw">set.seed</span>(<span class="dv">200413</span>)</a>
<a class="sourceLine" id="cb7-17" title="17"><span class="cf">for</span> (i <span class="cf">in</span> taskSubList[[taskId]]){</a>
<a class="sourceLine" id="cb7-18" title="18"><span class="co">#for (i in 1:nrow(matTable)) {</span></a>
<a class="sourceLine" id="cb7-19" title="19">  regionName=matTable<span class="op">$</span>region[i]</a>
<a class="sourceLine" id="cb7-20" title="20">  sampleName=matTable<span class="op">$</span>sample[i]</a>
<a class="sourceLine" id="cb7-21" title="21">  outFileBase=<span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)</a>
<a class="sourceLine" id="cb7-22" title="22">  <span class="kw">print</span>(<span class="kw">paste</span>(<span class="st">&quot;Clustering&quot;</span>, outFileBase))</a>
<a class="sourceLine" id="cb7-23" title="23">  dataMatrix&lt;-<span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, matTable<span class="op">$</span>filename[i], </a>
<a class="sourceLine" id="cb7-24" title="24">                              <span class="dt">package =</span> <span class="st">&quot;EMclassifieR&quot;</span>, <span class="dt">mustWork=</span><span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb7-25" title="25">  <span class="kw">dim</span>(dataMatrix)</a>
<a class="sourceLine" id="cb7-26" title="26">  dataMatrix&lt;-<span class="kw">removeAllNArows</span>(dataMatrix)</a>
<a class="sourceLine" id="cb7-27" title="27">  <span class="kw">dim</span>(dataMatrix)</a>
<a class="sourceLine" id="cb7-28" title="28">  </a>
<a class="sourceLine" id="cb7-29" title="29">  <span class="kw">tryCatch</span>( </a>
<a class="sourceLine" id="cb7-30" title="30">    {</a>
<a class="sourceLine" id="cb7-31" title="31">      <span class="kw">runEMrangeClassNum</span>(dataMatrix, k_range, convergenceError, maxIterations,</a>
<a class="sourceLine" id="cb7-32" title="32">                     <span class="dt">repeats=</span>numRepeats, <span class="dt">outPath=</span>outPath, <span class="dt">xRange=</span>xRange, </a>
<a class="sourceLine" id="cb7-33" title="33">                     <span class="dt">outFileBase=</span><span class="kw">paste</span>(sampleName, regionName, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>),</a>
<a class="sourceLine" id="cb7-34" title="34">                     <span class="dt">doIndividualPlots=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-35" title="35">  </a>
<a class="sourceLine" id="cb7-36" title="36">      <span class="kw">plotClusteringMetrics</span>(dataMatrix, k_range, maxB, convergenceError,</a>
<a class="sourceLine" id="cb7-37" title="37">                       maxIterations, outPath, outFileBase, nThreads,</a>
<a class="sourceLine" id="cb7-38" title="38">                       setSeed)</a>
<a class="sourceLine" id="cb7-39" title="39">    },</a>
<a class="sourceLine" id="cb7-40" title="40">    <span class="dt">error=</span><span class="cf">function</span>(e){<span class="st">&quot;Matrix not valid&quot;</span>}</a>
<a class="sourceLine" id="cb7-41" title="41">  )</a>
<a class="sourceLine" id="cb7-42" title="42">}</a></code></pre></div>
<pre><code>## [1] &quot;Clustering dS02-182_WBGene00015947&quot;
## [1] &quot;numClasses: 2&quot;
## [1] &quot;converged: TRUE iterations:  26&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  28&quot;
## [1] &quot;converged: TRUE iterations:  24&quot;
## [1] &quot;converged: TRUE iterations:  30&quot;
## [1] &quot;converged: TRUE iterations:  19&quot;
## [1] &quot;converged: TRUE iterations:  19&quot;
## [1] &quot;converged: TRUE iterations:  22&quot;
## [1] &quot;converged: TRUE iterations:  19&quot;
## [1] &quot;converged: TRUE iterations:  27&quot;
## [1] &quot;converged: TRUE iterations:  20&quot;</code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 3&quot;
## [1] &quot;converged: TRUE iterations:  23&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  32&quot;
## [1] &quot;converged: TRUE iterations:  26&quot;
## [1] &quot;converged: TRUE iterations:  33&quot;
## [1] &quot;converged: TRUE iterations:  25&quot;
## [1] &quot;converged: TRUE iterations:  27&quot;
## [1] &quot;converged: TRUE iterations:  28&quot;
## [1] &quot;converged: TRUE iterations:  39&quot;
## [1] &quot;converged: TRUE iterations:  33&quot;
## [1] &quot;converged: TRUE iterations:  33&quot;</code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 4&quot;
## [1] &quot;converged: TRUE iterations:  48&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  13&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 3</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  28&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 2</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  33&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 3</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  18&quot;
## [1] &quot;converged: TRUE iterations:  54&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  38&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  24&quot;
## [1] &quot;converged: TRUE iterations:  41&quot;
## [1] &quot;converged: TRUE iterations:  24&quot;</code></pre>
<pre><code>## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 5&quot;
## [1] &quot;converged: TRUE iterations:  43&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  44&quot;
## [1] &quot;converged: TRUE iterations:  36&quot;
## [1] &quot;converged: TRUE iterations:  28&quot;
## [1] &quot;converged: TRUE iterations:  55&quot;
## [1] &quot;converged: TRUE iterations:  47&quot;
## [1] &quot;converged: TRUE iterations:  22&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  24&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 2</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  40&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  73&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4
## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 6&quot;
## [1] &quot;converged: TRUE iterations:  24&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  50&quot;
## [1] &quot;converged: TRUE iterations:  33&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  42&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 5</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  35&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 6</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  28&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  40&quot;
## [1] &quot;converged: TRUE iterations:  35&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 6, 3</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  15&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  17&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 6
## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 7&quot;
## [1] &quot;converged: TRUE iterations:  36&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  34&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 3</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  37&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 7</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  37&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 5</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  27&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  17&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 7</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  33&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 2</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  86&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 6</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  40&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  31&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 4
## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
<pre><code>## [1] &quot;numClasses: 8&quot;
## [1] &quot;converged: TRUE iterations:  28&quot;</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  26&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 7</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  16&quot;
## [1] &quot;converged: TRUE iterations:  52&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 8</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  14&quot;
## [1] &quot;converged: TRUE iterations:  17&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  33&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 7</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  22&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 1</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  32&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 5, 6</code></pre>
<pre><code>## [1] &quot;converged: TRUE iterations:  27&quot;</code></pre>
<pre><code>## The following `from` values were not present in `x`: 6, 3
## `geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<pre><code>## Warning in allClassMeans[numClasses] &lt;- runEMrepeats(dataMatrix, numClasses, :
## number of items to replace is not a multiple of replacement length</code></pre>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
